File: golang/container/disjoint-set.go
package container

// 并查集
// - 带高度和路径压缩
// - n 从 0 开始。 1..n 需要+1
type DisjointSet struct {
	P    []int // 父节点
	Rank []int // 树高度
}

func InitDisjointSet(n int) *DisjointSet {
	d := &DisjointSet{}
	d.P = make([]int, n)
	d.Rank = make([]int, n)
	for i := range n {
		d.P[i] = i
	}
	return d
}

func (d *DisjointSet) InSame(u, v int) bool {
	return d.FindSet(u) == d.FindSet(v)
}

func (d *DisjointSet) Join(u, v int) {
	if d.FindSet(u) != d.FindSet(v) {
		d.link(d.FindSet(u), d.FindSet(v))
	}
}

func (d *DisjointSet) FindSet(u int) int {
	if u != d.P[u] {
		d.P[u] = d.FindSet(d.P[u])
	}
	return d.P[u]
}

func (d *DisjointSet) link(u, v int) {
	if d.Rank[u] > d.Rank[v] {
		d.P[v] = u
	} else {
		d.P[u] = v
		if d.Rank[u] == d.Rank[v] {
			d.Rank[u] = d.Rank[v] + 1
		}
	}
}

File: python/container/disjointset.py
"""
并查集
- 带高度和路径压缩
- n 从 0 开始
"""


class DisjointSet:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def in_same(self, u: int, v: int) -> bool:
        return self.find(u) == self.find(v)

    def union(self, u: int, v: int) -> bool:
        """
        返回值表示是否发生了合并（原来不在一个集合中）
        """
        if self.find(u) != self.find(v):
            self._merge(self.find(u), self.find(v))
            return True
        return False

    def find(self, u: int) -> int:
        if u != self.parent[u]:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def _merge(self, u: int, v: int) -> None:
        if self.rank[u] > self.rank[v]:
            self.parent[v] = u
        else:
            self.parent[u] = v
            if self.rank[u] == self.rank[v]:
                self.rank[u] = self.rank[v] + 1
